### DISCUSSION TITLE
Shader performance issue

### DISCUSSION THREAD

**Discussion by io-error** [cables-gl/cables_docs#295] [Answers & Questions]
Hello,

I have some issues with this shader :https://cables.gl/view/fXoFQo

Shaders and visual FX receive parameters created by FFT analysis (user defined bass,mid,high frequencies)
The problem is that the more the patch is running, the bigger the latency increases.

Do you think it is directly related to the sound analysis node?or the patch needs to be optimized?
and is there a way to monitor the CPU/GPU consumption of each node? 

thank you for your help

**Comment by steam0r** [comment]
hey, there is quite some documentation on debugging and optimizing your patches:

https://cables.gl/docs/9_performance_optimization/performance

for checking individual ops, basically put a performance op after MainLoop and then connect/disconnect and see the changes...

**Comment by io-error** [comment]
I already used the performance node but i just have around 20 ms /30 fps
It doesnt seem to be related to this
are there known performance issues with the audio analyzer?

**Reply by steam0r** [reply]
hm, i would not know of any. but looking at that massive patch with a lot of customshaders, midi, fft, postprocessing, ... i doubt it really is the audio. sorry, can't debug this community patch for you. would be great if you somehow could create really simple example that pinpoints the issue...this way maybe others can also assist.

**Comment by io-error** [comment]
im currently trying to externalize the sound analysis part to see if its the reason

**Comment by io-error** [comment]
so i removed the sound analysis part and used instead a python script to convert FFT to midi frames.
I got far better performance, and also did not experience the former problem of the latency drifting over time
As i prefer using sound analysis+fft embedded into cables, are there any known ways to improve its performance??

**Reply by tastez** [reply]
pretty hard to guess where the problem lies, do you have a simple patch to look at that shows that FFT is the issue? Also just to be clear - what do you mean by midi frames? are you using a midi file also? Anyway, just keep taking out things until it runs as smooth as you want is my approach. Then rebuild the stuff you can't live without in a simpler way :)
